File: projects/tax-id/src/lib/countries/senegal.ts
import { TaxIdValidationResult } from '../models';

const COFI_PATTERN = /^[012][ABCDEFGHJKLMNPQRSTUVWZ]\d$/;
const CHECKSUM_WEIGHTS = [1, 2, 1, 2, 1, 2, 1, 2, 1] as const;

export function validateSenegaleseTaxId(value: unknown): TaxIdValidationResult {
  const normalizedValue = normalizeSenegaleseTaxId(value);
  const base = { country: 'SN', normalizedValue } as const;

  if (!normalizedValue) return { ...base, valid: false, error: 'empty' };

  const hasCofi = normalizedValue.length === 10 || normalizedValue.length === 12;
  const ninea = hasCofi ? normalizedValue.slice(0, -3) : normalizedValue;
  const cofi = hasCofi ? normalizedValue.slice(-3) : '';

  if (ninea.length !== 7 && ninea.length !== 9) {
    return { ...base, valid: false, error: 'invalid_length' };
  }
  if (!/^\d+$/.test(ninea) || (cofi && !COFI_PATTERN.test(cofi))) {
    return { ...base, valid: false, error: 'invalid_format' };
  }

  const paddedNinea = ninea.padStart(9, '0');
  const checksum = [...paddedNinea].reduce(
    (sum, digit, index) => sum + Number(digit) * CHECKSUM_WEIGHTS[index],
    0,
  );

  if (checksum % 10 !== 0) {
    return { ...base, valid: false, error: 'invalid_checksum' };
  }

  return { ...base, valid: true, validationLevel: 'checksum' };
}

function normalizeSenegaleseTaxId(value: unknown): string {
  if (typeof value !== 'string' && typeof value !== 'number') return '';
  return String(value).trim().replace(/[\s\-/ ,]+/g, '').toUpperCase();
}

File: packages/dotnet/NationalIdentifiers.Core/Countries/Senegal.cs
using System.Text.RegularExpressions;
namespace NationalIdentifiers.Core.Countries;
internal static class Senegal
{
    private static readonly Regex CofiPattern = new(@"^[012][ABCDEFGHJKLMNPQRSTUVWZ]\d$", RegexOptions.Compiled);
    private static readonly int[] ChecksumWeights = { 1,2,1,2,1,2,1,2,1 };

    internal static ValidationResult Validate(object? value)
    {
        var n = value is string s
            ? Regex.Replace(s.Trim().ToUpperInvariant(), @"[\s\-/,]+", "")
            : string.Empty;
        if (string.IsNullOrEmpty(n)) return ValidationResult.Fail("SN", n, ValidationErrorCode.Empty);
        bool hasCofi = n.Length == 10 || n.Length == 12;
        string ninea = hasCofi ? n[..^3] : n;
        string cofi  = hasCofi ? n[^3..] : string.Empty;
        if (ninea.Length != 7 && ninea.Length != 9)
            return ValidationResult.Fail("SN", n, ValidationErrorCode.InvalidLength);
        if (!Regex.IsMatch(ninea, @"^\d+$") || (cofi.Length > 0 && !CofiPattern.IsMatch(cofi)))
            return ValidationResult.Fail("SN", n, ValidationErrorCode.InvalidFormat);
        string padded = ninea.PadLeft(9, '0');
        int checksum = padded.Select((digit, index) => (digit - '0') * ChecksumWeights[index]).Sum();
        if (checksum % 10 != 0)
            return ValidationResult.Fail("SN", n, ValidationErrorCode.InvalidChecksum);
        return ValidationResult.Ok("SN", n, ValidationLevel.Checksum);
    }
}
